File: 207-CourseSchedule/207-CourseSchedule.cpp
// Last updated: 02/08/2025, 18:49:20
class Solution {
public:
    bool canFinish(int numCourses, vector<vector<int>>& prerequisites) {
        // topological sort
        vector<vector<int>> adj(numCourses, vector<int>());
        for(auto it: prerequisites){
            adj[it[0]].push_back(it[1]);
        }
        // int indegree[numCourses] = {0};
        vector<int> indegree(numCourses, 0);
        for(int i = 0; i<numCourses; i++){
            for(auto it: adj[i]){
                indegree[it]++;
            }
        }
        queue<int> q;
        vector<int> topo;
        for(int i = 0 ; i< numCourses; i++){
            if(indegree[i] == 0){
                q.push(i);
            }
        }
        while(!q.empty()){
            int node = q.front();
            q.pop();
            topo.push_back(node);
            for(auto it: adj[node]){
                indegree[it]--;
                if(indegree[it] == 0){
                    q.push(it);
                }
            }
        }
        if(topo.size() == numCourses){
            return true;
        }
        return false;

    }
};

File: 207-CourseSchedule/207-CourseSchedule.py
# Last updated: 01/10/2025, 00:13:17
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:

        graph = {i: [] for i in range(numCourses)}
        for course, prereq in prerequisites:
            graph[prereq].append(course)
        state = [0]*numCourses
        def dfs(node):
            if state[node] == 1:
                return False
            if state[node] == 2:
                return True
            state[node] = 1
            for nei in graph[node]:
                if not dfs(nei):
                    return False
            
            state[node] = 2
            return True
        
        for course in range(numCourses):
            if state[course] == 0:
                if not dfs(course):
                    return False
        return True
        